hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ServiceInfoProvider.java

```diff
@@ -0,0 +1,145 @@
+import org.slf4j.Logger;
+
+import java.io.IOException;
+import java.security.cert.X509Certificate;
+import java.time.Instant;
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static java.util.Collections.emptyList;
+import static org.slf4j.LoggerFactory.getLogger;
+
+/**
+ * A helper class to handle the construction of
+ * {@link org.apache.hadoop.ozone.om.helpers.ServiceInfoEx} objects,
+ * and refresh of the things that are necessary for constructing it.
+ */
+public final class ServiceInfoProvider {
+
+  private static final Logger LOG = getLogger(ServiceInfoProvider.class);
+
+  private final OzoneManagerProtocol om;
+  private final CertificateClient certClient;
+
+  private String caCertPem;
+  private List<String> caCertPemList;
+
+  public ServiceInfoProvider(
+      OzoneConfiguration config,
+      OzoneManagerProtocol om,
+      CertificateClient certClient
+  ) throws IOException {
+    this(config, om, certClient, false);
+  }
+
+  public ServiceInfoProvider(
+      OzoneConfiguration config,
+      OzoneManagerProtocol om,
+      CertificateClient certClient,
+      boolean skipInitializationForTesting
+  ) throws IOException {
+    this.om = om;
+    if (new SecurityConfig(config).isSecurityEnabled()
+        && !skipInitializationForTesting) {
+      this.certClient = certClient;
+      Set<X509Certificate> certs = getCACertificates();
+      caCertPem = toPEMEncodedString(newestOf(certs));
+      caCertPemList = toPEMEncodedStrings(certs);
+      this.certClient.registerRootCARotationListener(onRootCAChange());
+    } else {
+      this.certClient = null;
+      caCertPem = null;
+      caCertPemList = emptyList();
+    }
+  }
+
+  private Function<List<X509Certificate>, CompletableFuture<Void>>
+      onRootCAChange() {
+    return certs -> {
+      CompletableFuture<Void> returnedFuture = new CompletableFuture<>();
+      try {
+        caCertPem = toPEMEncodedString(newestOf(certs));
+        caCertPemList = toPEMEncodedStrings(certs);
+        returnedFuture.complete(null);
+      } catch (Exception e) {
+        LOG.info("Unable to refresh cached PEM formatted CA " +
+            "certificates.", e);
+        returnedFuture.completeExceptionally(e);
+      }
+      return returnedFuture;
+    };
+  }
+
+  public ServiceInfoEx provide() throws IOException {
+    return new ServiceInfoEx(om.getServiceList(), caCertPem, caCertPemList);
+  }
+
+  private Set<X509Certificate> getCACertificates() {
+    return !certClient.getAllRootCaCerts().isEmpty()
+        ? certClient.getAllRootCaCerts()
+        : certClient.getAllCaCerts();
+  }
+
+  private X509Certificate newestOf(Collection<X509Certificate> certs) {
+    return certs.stream()
+        .max((c1, c2) -> expiry(c1).compareTo(expiry(c2)))
```

Comment:
Please use the inline statement directly, instead of calling a one statement function.

Reply:
JIT would inline the method call anyways, and probably it would do more. Though this note was useful because it made me check what I was doing, and realize that I have a way better solution to solve the readability problem I wanted to solve with the expiry method.